src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs

```diff
@@ -14,12 +14,21 @@ namespace Microsoft.PowerFx.Functions
     // Direct ports from JScript. 
     internal static partial class Library
     {
+        /// <summary>
+        /// Get a service from the <paramref name="serviceProvider"/>,
+        /// Returns null if not present.
+        /// </summary>
+        public static T GetService<T>(IServiceProvider serviceProvider)
+        {
+            return (T)serviceProvider.GetService(typeof(T));
+        }
+
         private static readonly IRandomService _defaultRandService = new DefaultRandomService();
 
         // Support for aggregators. Helpers to ensure that Scalar and Tabular behave the same.
         private interface IAggregator
         {
-            void Apply(FormulaValue value);
+            void Apply(FormulaValue value, IServiceProvider serviceProvider);
```

Comment:
>, IServiceProvider serviceProvider [](http://example.com/codeflow?start=41&length=34)

The parameter here means every call to the aggregator will have to repeat the lookup. 

Aggregators are short lived and allocated here .
So instead, can we pass the parameter to the ctor and save as a field?

And then IAggregator is unchanged.

Reply:
That's a great idea! Thanks!